mm2src/coins/nft/trading_proto_v2/mod.rs

```diff
@@ -0,0 +1,565 @@
+    /// and encodes the input parameters required for the blockchain transaction.
+    fn prepare_spend_nft_maker_v2_data(
+        &self,
+        contract_type: ContractType,
+        args: &SpendNftMakerPaymentArgs<'_, Self>,
+        decoded: Vec<Token>,
+        htlc_params: Vec<Token>,
+        state: U256,
+    ) -> Result<Vec<u8>, PrepareTxDataError> {
+        let spend_func = match contract_type {
+            ContractType::Erc1155 => NFT_SWAP_CONTRACT.function("spendErc1155MakerPayment")?,
+            ContractType::Erc721 => NFT_SWAP_CONTRACT.function("spendErc721MakerPayment")?,
+        };
+
+        if state != U256::from(MakerPaymentStateV2::PaymentSent as u8) {
+            return Err(PrepareTxDataError::Internal(ERRL!(
+                "Payment {:?} state is not PAYMENT_STATE_SENT, got {}",
+                args.maker_payment_tx,
+                state
+            )));
+        }
+
+        let input_tokens = match contract_type {
+            ContractType::Erc1155 => vec![
+                htlc_params[0].clone(), // swap_id
+                Token::Address(args.maker_payment_tx.sender()),
+                Token::FixedBytes(args.taker_secret_hash.to_vec()),
+                Token::FixedBytes(args.maker_secret.to_vec()),
+                htlc_params[2].clone(), // tokenAddress
+                decoded[2].clone(),     // tokenId
+                decoded[3].clone(),     // amount
+            ],
+            ContractType::Erc721 => vec![
+                htlc_params[0].clone(), // swap_id
+                Token::Address(args.maker_payment_tx.sender()),
+                Token::FixedBytes(args.taker_secret_hash.to_vec()),
+                Token::FixedBytes(args.maker_secret.to_vec()),
+                htlc_params[2].clone(), // tokenAddress
+                decoded[2].clone(),     // tokenId
+            ],
+        };
+
+        let data = spend_func.encode_input(&input_tokens)?;
+        Ok(data)
+    }
+
+    async fn status_and_htlc_params_from_tx_data(
+        &self,
+        swap_contract: Address,
+        contract: &Contract,
+        decoded_data: &[Token],
+        index: usize,
+        state_type: StateType,
+    ) -> Result<(U256, Vec<Token>), PaymentStatusErr> {
+        if let Some(Token::Bytes(data_bytes)) = decoded_data.get(index) {
+            if let Ok(htlc_params) = ethabi::decode(htlc_params(), data_bytes) {
+                let state = self
+                    .payment_status_v2(swap_contract, htlc_params[0].clone(), contract, state_type)
+                    .await?;
+                Ok((state, htlc_params))
+            } else {
+                Err(PaymentStatusErr::TxDeserializationError(ERRL!(
+                    "Failed to decode HTLCParams from data_bytes"
+                )))
+            }
+        } else {
+            Err(PaymentStatusErr::TxDeserializationError(ERRL!(
+                "Failed to decode HTLCParams from data_bytes"
+            )))
+        }
```

Comment:
Please avoid nesting here e.g.
```rust
    let data_bytes = match decoded_data.get(index) {
        Some(Token::Bytes(data_bytes)) => data_bytes,
        _ => return Err(PaymentStatusErr::TxDeserializationError(ERRL!("Failed to decode HTLCParams from data_bytes"))),
    };

    let htlc_params = match ethabi::decode(htlc_params(), data_bytes) {
        Ok(htlc_params) => htlc_params,
        Err(_) => return Err(PaymentStatusErr::TxDeserializationError(ERRL!("Failed to decode HTLCParams from data_bytes"))),
    };

    let state = self
        .payment_status_v2(swap_contract, htlc_params[0].clone(), contract, state_type)
        .await?;

    Ok((state, htlc_params))
```

Reply:
fixed here with some other places https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/977cf40af09bfb0046b306565dfdf68b99a79929

Reply:
Thanks, like how you did the checks vec :)